buildSrc/src/main/groovy/com.google.api-ads.nexus-publish.gradle

```diff
@@ -0,0 +1,274 @@
+        this.project = project
+        this.authHeaders = new AuthHeaders(
+                project.properties.get("sonatypeUser"),
+                project.properties.get("sonatypePassword"))
+    }
+
+    /**
+     * Provides the Sonatype repository URL to use for deploying releases.
+     *
+     * Attempts to create a named staging repository to avoid using Nexus auto-staging if
+     * possible. If not possible to create a named staging repository, falls back to auto-staging.
+     *
+     * This is motivated by the observation that auto-staging's heuristic for which artifacts
+     * belong in the same repository is often confused by Gradle's publish mechanism, which
+     * appears to create a new TCP connection for each artifact. This also appears to be flaky
+     * based on the sonatype server load.
+     */
+    def discoverSonatypeRepositoryUrl() {
+        try {
+            // Ensures that we don't create spurious repositories if we're not running a release.
+            if (project.properties.containsKey("release")) {
+                // Gradle may reuse the demon (hence its classloader) so this uses a map to store
+                // the staging repo for each build.
+                return STAGING_URL_CACHE.computeIfAbsent(
+                        project.rootProject, { project -> this.createStagingRepository() })
+            } else {
+                logger.debug("Release not enabled, defaulting to auto-staging")
+                return AUTO_STAGING_URL
+            }
+        } catch (Exception ex) {
+            // Handles issues by falling back to auto-staging.
+            logger.warn("Failed to create new staging repo. Attempting to deploy with " +
+                    "auto-staging. The release *may* complete as planned. You can proceed by" +
+                    " making sure that all release artifacts are present in the auto-staging " +
+                    "repository. Please investigate why staging failed to create.", ex)
+            return AUTO_STAGING_URL
+        }
+    }
+
+    /** Creates a staging repository on Sonatype. */
+    private def createStagingRepository() {
+        logger.info("Creating staging repository")
+        def stagingProfileUrls = getStagingProfileUrls()
+        if (stagingProfileUrls.size() == 0) {
+            throw new GradleException("No staging profiles found")
+        } else if (stagingProfileUrls.size() > 1) {
+            logger.warn("Found multiple staging profiles, using the first one")
+        }
+        def stagingStartUrl = new GenericUrl(stagingProfileUrls[0] + "/start")
+        def content = new PromoteRequestXmlContent(
+                NAMESPACE_DICTIONARY, generateStagingRepoDescription())
+        def response = REQUEST_FACTORY
+                .buildPostRequest(stagingStartUrl, content)
+                .setHeaders(authHeaders)
+                .execute()
+        // Checks the response and returns the repository URL, falling back to auto-staging if
```

Comment:
nit: comment on fallback is referring to the caller behavior. This should say it throws an exception.